Bug 983096: Batch operations

mathjazz opened the pull request:
This is not ready for a review yet.

mathjazz commented:
Batch editing brings:
* Ability for to select entities in entity list and perform actions on them in a batch editor (approve all, delete all, find & replace). It's only exposed to privileged users (Translators).
* Colored checkboxes as status icons in entitylist (multi select)
* Colored radio buttons as status icons in history tab (single select)
* Ability to select multiple entities by holding Shift
* Ability to select all entities (even if not in entity list due to progressive loading)

This PR also contains commits with some other bugfixes and improvements:
* 9db6938 Remove unused old parts menu CSS
* ef5eedb No tab stop on hidden download/upload forms
* 0e0f7bb Unify alignment of search field and editor topbar in translate view
* 992e028 & a166edb Prevent notifications from covering the Next button
* 03608fa No need for entity key in entity list 
* a830b0d Expose /manage to managers, not translators
* 0317fec Join updateTranslation() and updateEntityUI()

**Note**
While this is ready for a review (and branch out for other frontend facing bugs), there's one open issue left I'm still working on and I'll fix in a separate commit. If lots (1000s) of entities are batch-edited, request will most probably timeout on production. It takes over 20 seconds locally for all entities selected in the currently biggest project (Firefox Aurora).

There are 2 slow parts, both related to the DB:
1. Getting "active" (approved or latest) translations for each entity (on which batch operations are ran): https://github.com/mozilla/pontoon/pull/369/files#diff-d964cbd7ee26bf4860da2c98511ccb58R456
2. (Related to 1) Updating stats for all changed TranslatedResources: https://github.com/mozilla/pontoon/pull/369/files#diff-d964cbd7ee26bf4860da2c98511ccb58R527

We'll probably have to change model to make it easier to get currently active translations, because we need that operation to be quick on multiple occasions: when getting entities for the entity list, when calculating stats and here - when doing batch operations. It should probably be a separate bug, because it might get quite heavy, and in the meantime we could split big requests in multiple requests (e.g. 1 per 1000 entities).

@jotes r?

jotes commented:
A first round of questions/thoughts:
* delete of selected translation causes reload of  -> it would be a good thing to notify user what happened
and how many items were removed, because user could not catch that status was displayed on the button right before reload.
* progressbar display invalid counts for entities/doesn't update after mass action.
* if you click on the label of entity by the mistake, previously selected entities are unselected again and panel with mass-action vanishes.
* i guess that "reversible" batch operations will be our next feature ;-)
* we could also add a prompt("Are you sure?") for the mass delete.
* alternatively we could add a button that would sync selected translations with their latest states from the repository.
* also, if user for some reason selects all entities, we could do this the way gmail handles selecting all e-mails (that wouldn't require us to load all entities to client)

# nits/questions
* select all could be below a list of entities and the status of selected items and buttons for the actions? It's just an idea, i'm not against the current state of this feature.

jotes commented on pontoon/base/forms.py:
```diff
@@ -7,13 +7,13 @@
-    slug = forms.CharField()
-    code = forms.CharField()
-    part = forms.CharField()
+    slug = forms.CharField(widget=forms.TextInput(attrs={'tabindex': '-1'}))
```
We could create our custom fields to avoid duplication of the code.

jotes commented on pontoon/base/forms.py:
```diff
@@ -7,13 +7,13 @@
 
 class UploadFileForm(DownloadFileForm):
-    uploadfile = forms.FileField()
+    uploadfile = forms.FileField(widget=forms.FileInput(attrs={'tabindex': '-1'}))
```
same here.

jotes commented on pontoon/base/static/css/translate.css:
```diff
@@ -1299,3 +1354,116 @@ p#sign-in-required > a#sidebar-signin {
 }
+
+
+
```
too many newlines?

mathjazz replied:
:-) Probably, but I'll keep this consistent with the rest of the file.

jotes commented on pontoon/base/static/js/translate.js:
```diff
@@ -271,6 +282,24 @@ var Pontoon = (function (my) {
+
+      $('#batch')
+        .find('button').removeClass('loading loaded')
+          .find('.message').html('');
```
I would match the indentation with the previous statement.

mathjazz replied:
Actually, if you look at the first block, there's `.end()` at the end, which means we jump a level higher. So `.variant` and `.selected-count` are siblings.

In the 2nd block, `.message` is a child of the `button`, hence the indent.

jotes commented on pontoon/base/static/js/translate.js:
```diff
@@ -271,6 +282,24 @@ var Pontoon = (function (my) {
     /*
+     * Open batch editor
+     */
+    openBatchEditor: function () {
```
Idea for the next refactorings: I guess we could create an separate object called an Editor and add proper methods like setSelectedCount. At some point it's a really hard to figure what selectors are really meaning for the application.
However, that's not something we should do in this pull request.

mathjazz replied:
Agreed. Also, there are many global variable we're assigning to Pontoon object and it's hard to keep track of them. We should spread them among such objects and maybe keep them in ObjectName.data instead of the top level, so they're easier to manage.

jotes commented on pontoon/base/static/js/translate.js:
```diff
@@ -1076,6 +1166,133 @@ var Pontoon = (function (my) {
+          error: function(error) {
+            message = 'Oops, something went wrong';
+          },
+          complete: function() {
```
It's a philosophical thing but maybe it's better to handle success/error in the complete() ? it would be easier to track (i mean read) what happens with the message.

mathjazz replied:
Probably true, but I'm leaving it for later. In case we decide to do it, we should do it everywhere.

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
     }, safe=False)
 
 
+@login_required(redirect_field_name='', login_url='/403')
```
We could use functools partial to create our version of login required.

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+                'error': 'Empty translations not allowed',
+            })
+
+        # Find translations
```
I would move this whole part into TranslationManager, i guess it will be an easier to test later.

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+            translation.pk = None  # Create new translation
+            translation.string = translation.string.replace(find, replace)
+            translation.user = translation.approved_user = request.user
+            translation.date = translation.approved_date = timezone.now()
```
Maybe it's a little premature optimalization but we could call this method once per batch.

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+        # Empty translations produced by replace might not be always allowed
+        forbidden = (
+            translations.filter(string=find)
+            .exclude(entity__resource__format__in=['properties', 'ini', 'dtd'])
```
We could move is_asymmetric property into database, it would remove redundancies like this.

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+            })
+
+        # Find translations
+        translations = translations.filter(string__contains=find)
```
I guess we could use defer() here, because initialization of instances of models cost a lot of cpu/memory. Alternatively  we could use values() and pack these values into Translation() objects. We could test which way will be faster.

mathjazz replied:
Thanks for the suggestion, I'll investigate!

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+
+        # Find translations
+        translations = translations.filter(string__contains=find)
+        translatedresources = translations.translated_resources()
```
Am i right that we could check if translations.count() == 0 and return JsonResponse({'count': 0}) here?

jotes commented on pontoon/base/views.py:
```diff
@@ -427,6 +433,106 @@ def entities(request):
+@require_POST
+@require_AJAX
+@transaction.atomic
+def batch_edit_translations(request):
```
I'm not sure if it wouldn't be such bad idea to consider moving some parts of this view into celery tasks because there's a chance that parts like calculation of stats will take a while.

mathjazz replied:
I'll see what I can do with stats, celery is also an option, but probably not the most user friendly one.

jotes commented on pontoon/base/templates/translate.html:
```diff
@@ -218,101 +238,155 @@ <h3 class="loading"><div class="fa fa-refresh fa-spin"></div></h3>
-    </div>
+        <div class="review">
+          <h2>Review translations</h2>
+          <button id="approve-all">
```
I would create {% macro button(id, title) %}

jotes commented on pontoon/base/templates/translate.html:
```diff
@@ -218,101 +238,155 @@ <h3 class="loading"><div class="fa fa-refresh fa-spin"></div></h3>
+        {% endif %}
+      </menu>
+
+      <div id="helpers" class="tabs">
```
i guess we could refactor these tabs into jinja macros. However, maybe its a thing for the separate task.

jotes commented:
This is awesome feature :) i've pointed some small nits and written some ideas for the next things/improvements in future. Anyway, after i tested this locally - a whole feature works fine.
About the issues with the performance, i have only two ideas for now:
- we could created an another cache object with precalculated status of translation
- heavy calculations could be done in separate celery task.

I'll ping you if i'll figure out something better. Anyway, you have my r+wc for now.

mathjazz commented:
Thanks for the feedback!

>     delete of selected translation causes reload of -> it would be a good thing to notify user what happened and how many items were removed, because user could not catch that status was displayed on the button right before reload.

Yeah, we actually reload after any batch operation, because that was the easiest way not to break anything. I'll try to be smarter about it.

>     progressbar display invalid counts for entities/doesn't update after mass action.

Thanks, fixed locally, will push once you're done with the review.

>     if you click on the label of entity by the mistake, previously selected entities are unselected again and panel with mass-action vanishes.

True, but that's also the only way to clear selection in the single-column layout (not sure there's a good spot for the clear selection button). I've extended the entity checkbox selection area to window border in the coming commit, so you can move your mouse to the left of your screen and just move it vertically to select entities. With filters and keyboard shortcuts it should also be relatively easy to select entities.

>     i guess that "reversible" batch operations will be our next feature ;-)

Yeah. Should be fun. ;)

>     we could also add a prompt("Are you sure?") for the mass delete.

True, but then again we have this warning introductory text on top. If we simply show the prompt every time, we might warn users before the change and then it turns out no changes actually happen. So we'd have to be smarter, make the request, see if there are any changes pending, show the prompt and then make the request again. Maybe we could add this in the next iteration.

>     alternatively we could add a button that would sync selected translations with their latest states from the repository.

Good idea. It's just that we'd need to make big changes to sync to make this work.

>     also, if user for some reason selects all entities, we could do this the way gmail handles selecting all e-mails (that wouldn't require us to load all entities to client)

Actually we already do this. We select everything, without loading entities on the client side.

>     select all could be below a list of entities and the status of selected items and buttons for the actions? It's just an idea, i'm not against the current state of this feature.

Yeah I'm not entirely happy with the UI. We have something like this in the single-column layout (check out the homepage). Later on we'll probably add more features (buttons), so we'll need a full panel anyways.

mathjazz commented:
Thanks for the very valuable input, @jotes! I've added 3 commits addressing your comments.

**TODO**
* [ ] Do not reload the UI after batch changes are made.
* [ ] Prevent timing out on huge requests.

mathjazz commented:
@jotes In the last 5 commits I've addressed your comments and fixed a bug that prevented batch updates to sync (ed126eb). This is ready for a 2nd look and then I'll rebase.

Sadly I wasn't able to fix the problem with big requests. I'm thinking about filing a bug to move bigger requests to celery, but I'm not sure how the UX would look in this case. Do you have any experience with that? Until then, we could set a hard limit of e.g. max. 3000 strings to be edited at once.

jotes commented:
I had some experiences with celery in my previous projects.
For the starters: Javascript sends an request to server. Django view creates an async task and returns an id of the task result to the client.
Client starts periodically (alternatively via websockets) polling data from a separate resource (e.g. /task/status/?id=<id task). This resource should return a current status (or the result) of task.
If task succeeded, the /task/status resource should return the results. Javascript on the client-side can display this result in UI via some kind of a popup message/notification.

mathjazz commented:
Thanks again, I've implemented your comments and added an optimization for stats calculation.

I'll experiment a bit on stage and see where should I set the hard limit. Then we can rebase, merge, deploy and file a bug to move big requests to celery.

I'm also thinking about adding a model class for faster access to currently active translations, which is also needed when getting & filtering entity translations for entity list and when calculating stats. Something like this:

```
class ActiveTranslation(models.Model):
    entity = models.ForeignKey(Entity)
    locale = models.ForeignKey(Locale)
    translations = models.ManyToManyField(Translation)
```
